presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
```

Comment:
This normalization is needed because we want to effectively handle expressions like 
`JOIN ON NOT a.x = b.x`. Without normalization the join condition would be put in postInnerJoin conditions which is not what we want.

But you are are probably right that this normalization should happen in different place - as part of standard rewrite flow. Currently I cannot put it in `SimplifyExpressions` as JoinNode does not contain original join condition; only list of `EquiJoinClause`s. And we do not have access to Analysis object in `SimplifyExpressions` rewriter. And even if had it, we probably do not want change it there. Right?

Did your intention was to add original join condition to JoinNode? And then change `SimplifyExpression` to include normalization and also add handling of JoinNodes?